folio-spring-i18n/src/main/java/org/folio/spring/model/TranslationMap.java

```diff
@@ -0,0 +1,185 @@
+
+  @Getter
+  @Nullable
+  protected final TranslationMap fallback;
+
+  /**
+   * Create a TranslationMap for a given locale based off of a file with contents and a given fallback.
+   *
+   * @param locale the associated locale
+   * @param file the {@link TranslationFile TranslationFile} to infer quality from
+   * @param patterns the patterns to use
+   * @param fallback the TranslationMap to search if a given translation cannot be found
+   */
+  protected TranslationMap(
+    Locale locale,
+    TranslationFile file,
+    Map<String, String> patterns,
+    @Nullable TranslationMap fallback
+  ) {
+    this.locale = locale;
+    this.quality = TranslationMatchQuality.getQuality(locale, file);
+    this.file = file;
+    this.patterns = patterns;
+    this.fallback = fallback;
+  }
+
+  /**
+   * Create a TranslationMap for a given locale based off of a file with a given fallback.
+   *
+   * @param locale the associated locale
+   * @param file the {@link TranslationFile TranslationFile} to read from
+   * @param fallback the TranslationMap to search if a given translation cannot be found
+   */
+  public TranslationMap(
+    Locale locale,
+    TranslationFile file,
+    @Nullable TranslationMap fallback
+  ) {
+    this(locale, file, file.getMap(), fallback);
+  }
+
+  /**
+   * Create a <em>default</em> TranslationMap for a given locale based off of a file.
+   * This constructor should only be used for default translation maps; all others
+   * should use the other constructor and provide a fallback.
+   *
+   * @param locale the associated locale
+   * @param file the {@link TranslationFile TranslationFile} to read from
+   */
+  public TranslationMap(Locale locale, TranslationFile file) {
+    this(locale, file, null);
+  }
+
+  /**
+   * Create a new TranslationMap for a given locale.  This is primarily used when cloning
+   * existing TranslationMaps for use as fallbacks for another (for caching reasons, this
+   * makes more sense than re-reading files each time)
+   *
+   * @param newLocale the locale for the new TranslationMap to use
+   * @return the TranslationMap with {@code newLocale}
+   */
+  public TranslationMap withLocale(Locale newLocale) {
+    return new TranslationMap(
+      newLocale,
+      this.file,
+      this.patterns,
+      this.fallback
+    );
+  }
+
+  /**
+   * Get the ICU format string associated with a given key for this translation.
+   *
+   * @param key the key to lookup
+   * @return one of the following, with this precedence:
+   *   <ol>
+   *     <li>the ICU format string for this locale, if it is known</li>
+   *     <li>the ICU format string for a locale with a worse {@code TranslationMatchQuality},
+   *       such as one without a specific country, recursing through the {@code fallback}</li>
+   *     <li>the ICU format string for the server's default locale</li>
+   *     <li>the provided key</li>
+   *   </ol>
+   */
+  public String get(String key) {
```

Comment:
If the keys in the map are all lower-case, should this method also convert the param to lower-case?

Reply:
Most of the modules I've seen will use camelCase for their translation keys; There's nothing requiring the IDs to be lowercase, and looking at https://folio-snapshot.dev.folio.org/settings/developer/translations it seems like plenty aren't:

<img width="682" alt="image" src="https://github.com/folio-org/folio-spring-support/assets/8005215/b7eab53d-a29d-4dc2-9570-c7d8f9899a26">

Reply:
I was thinking of the cases where the map comes from `TranslationMap`, which converts them to lower-case. I think it should either be all lower-case or don't bother doing the conversion in `TranslationMap`, just to keep it consistent

Reply:
I'm not sure where you're thinking of — the only thing converted to lowercase is the locale (e.g. `en-US` to `en-us`), and that's done in `TranslationFile#getParts` and as part of `computeIfAbsent` in `TranslationService#getTranslation`